### PR TITLE
Exposing the pod manifest file/dir option for the node in Origin

### PR DESCRIPTION
**Pull request by abhgupta**
@deads2k PTAL

### CODE REVIEW

File: pkg/cmd/server/start/node_args.go [hunk 1]
```diff
@@ -37,6 +38,9 @@ type NodeArgs struct {
 	ImageFormatArgs    *ImageFormatArgs
 	KubeConnectionArgs *KubeConnectionArgs
 	CertArgs           *CertArgs
+
+	PodManifestPath    string
+	FileCheckFrequency time.Duration
```
Comment (liggitt) [inline pkg/cmd/server/start/node_args.go]: If we're not binding these and exposing them on the command line, we don't need to add them as args

File: pkg/cmd/server/start/node_args.go [hunk 2]
```diff
@@ -114,6 +118,9 @@ func (args NodeArgs) BuildSerializeableNodeConfig() (*configapi.NodeConfig, erro
 		DNSIP:     dnsIP,
 
 		MasterKubeConfig: admin.DefaultNodeKubeConfigFile(args.CertArgs.CertDir, args.NodeName),
+
+		PodManifestPath:    args.PodManifestPath,
+		FileCheckFrequency: args.FileCheckFrequency,
```
Comment (liggitt) [inline pkg/cmd/server/start/node_args.go]: default to `""` and `0`?

File: pkg/cmd/server/kubernetes/node.go [hunk 3]
```diff
@@ -126,7 +126,13 @@ func (c *NodeConfig) RunKubelet() {
 	recorder := record.NewBroadcaster().NewRecorder(kapi.EventSource{Component: "kubelet", Host: c.NodeHost})
 
 	cfg := kconfig.NewPodConfig(kconfig.PodConfigNotificationSnapshotAndUpdates, recorder)
-	kconfig.NewSourceApiserver(c.Client, c.NodeHost, cfg.Channel("api"))
+	kconfig.NewSourceApiserver(c.Client, c.NodeHost, cfg.Channel(kubelet.ApiserverSource))
+	// define manifest file source for pods
+	if _, err = os.Stat(c.PodManifestPath); err == nil {
```
Comment (liggitt) [inline pkg/cmd/server/kubernetes/node.go]: only do this if `len(PodManifestPath) != 0`

File: pkg/cmd/server/api/v1/types.go [hunk 4]
```diff
@@ -35,6 +37,13 @@ type NodeConfig struct {
 
 	// AllowDisabledDocker if true, the Kubelet will ignore errors from Docker.  This means that a node can start on a machine that doesn't have docker started.
 	AllowDisabledDocker bool `json:"allowDisabledDocker"`
+
+	// PodManifestPath specifies the path for the pod manifest file(s)
+	// The path could point to a single file or a directory that contains multiple manifest files
+	// This is used by the Kubelet to create pods on the node
+	PodManifestPath string `json:"podManifestPath"`
+	// FileCheckFrequency is the duration between checking manifest files for new data
+	FileCheckFrequency time.Duration `json:"fileCheckFrequency"`
```
Comment (liggitt) [inline pkg/cmd/server/api/v1/types.go]: can these be grouped, or named so it is obvious that `fileCheckFrequency` refers to the `podManifestPath`? In the serialized file, stuff gets written alphabetically

File: pkg/cmd/server/api/types.go [hunk 5]
```diff
@@ -36,6 +38,13 @@ type NodeConfig struct {
 
 	// AllowDisabledDocker if true, the Kubelet will ignore errors from Docker.  This means that a node can start on a machine that doesn't have docker started.
 	AllowDisabledDocker bool
+
+	// PodManifestPath specifies the path for the pod manifest file(s)
+	// The path could point to a single file or a directory that contains multiple manifest files
+	// This is used by the Kubelet to create pods on the node
+	PodManifestPath string
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: Any file paths have to be added to `GetNodeFileReferences`. Also add validation to `ValidateNodeConfig`

File: pkg/cmd/server/api/types.go [hunk 6]
```diff
@@ -36,6 +38,13 @@ type NodeConfig struct {
...
+	// FileCheckFrequency is the duration between checking manifest files for new data
+	FileCheckFrequency time.Duration
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: Add validation (can't be less than 0, probably shouldn't be more than once per second, etc). I would represent this as seconds, not as nanoseconds

**Comment by liggitt** [comment]
@deads2k what's the goal for create-node-config for stuff like this?

File: pkg/cmd/server/kubernetes/node_config.go [hunk 7]
```diff
@@ -54,6 +55,13 @@ type NodeConfig struct {
 	Client *client.Client
 	// A client to connect to Docker
 	DockerClient dockertools.DockerInterface
+
```
Comment (deads2k) [inline pkg/cmd/server/kubernetes/node_config.go]: I think there's enough shared now to warrant including the options (configapi.NodeConfig) directly.

**Comment by deads2k** [comment]
> what's the goal for create-node-config for stuff like this?

This looks to be completely optional.  Empty just means you don't have one of these.  If empty values are reasonable defaults, then I think we just let create-node-config leave the values empty.

**Comment by abhgupta** [comment]
@liggitt @deads2k made most of the changes - please take another look

File: pkg/cmd/server/kubernetes/node.go [hunk 8]
```diff
@@ -126,7 +126,17 @@ func (c *NodeConfig) RunKubelet() {
 	recorder := record.NewBroadcaster().NewRecorder(kapi.EventSource{Component: "kubelet", Host: c.NodeHost})
 
 	cfg := kconfig.NewPodConfig(kconfig.PodConfigNotificationSnapshotAndUpdates, recorder)
-	kconfig.NewSourceApiserver(c.Client, c.NodeHost, cfg.Channel("api"))
+	kconfig.NewSourceApiserver(c.Client, c.NodeHost, cfg.Channel(kubelet.ApiserverSource))
+	// define manifest file source for pods, if specified
+	if len(c.PodManifestPath) > 0 {
+		if _, err = os.Stat(c.PodManifestPath); err == nil {
```
Comment (liggitt) [inline pkg/cmd/server/kubernetes/node.go]: if there's an error here, should the startup fail?

File: pkg/cmd/server/api/validation/validation.go [hunk 9]
```diff
@@ -93,6 +93,15 @@ func ValidateRemoteConnectionInfo(remoteConnectionInfo api.RemoteConnectionInfo)
 	return allErrs
 }
 
+func ValidatePodManifestPath(path string, field string) fielderrors.ValidationErrorList {
+	allErrs := fielderrors.ValidationErrorList{}
+
+	allErrs = append(allErrs, ValidateFile(path, field)...)
+	// TODO: load and parse
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/validation.go]: note that the path can be a file or a directory

**Comment by liggitt** [comment]
a couple last comments, then LGTM. @deads2k?

File: pkg/cmd/server/api/types.go (hunk 5 above)
Comment (deads2k) [inline pkg/cmd/server/api/types.go]: These should be in their own struct, similar to https://github.com/openshift/origin/blob/master/pkg/cmd/server/api/types.go#L62.  That way they'll always be logically grouped regardless of serialization and if you make it a pointer it becomes clear that it is optional.

**Comment by deads2k** [comment]
one comment about the shape of the struct.  Otherwise, lgtm.

**Comment by abhgupta** [comment]
@liggitt @deads2k how about now?

File: pkg/cmd/server/api/types.go [hunk 10]
```diff
@@ -36,6 +36,10 @@ type NodeConfig struct {
 
 	// AllowDisabledDocker if true, the Kubelet will ignore errors from Docker.  This means that a node can start on a machine that doesn't have docker started.
 	AllowDisabledDocker bool
+
+	// PodManifestConfig holds the configuration for enabling the Kubelet to
+	// create pods based from a manifest file(s) placed locally on the node
+	PodManifestConfig *PodManifestConfig `json:"podManifestConfig"`
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: No JSON tags on internal version

File: pkg/cmd/server/api/validation/validation.go [hunk 11]
```diff
@@ -93,6 +93,21 @@ func ValidateRemoteConnectionInfo(remoteConnectionInfo api.RemoteConnectionInfo)
 	return allErrs
 }
 
+func ValidatePodManifestConfig(podManifestConfig *api.PodManifestConfig) fielderrors.ValidationErrorList {
+	allErrs := fielderrors.ValidationErrorList{}
+
+	if len(podManifestConfig.Path) > 0 {
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/validation.go]: If the podmanifestconfig is present, the path and check interval should be required

File: pkg/cmd/server/start/node_args.go [hunk 12]
```diff
@@ -114,6 +114,11 @@ func (args NodeArgs) BuildSerializeableNodeConfig() (*configapi.NodeConfig, erro
 		DNSIP:     dnsIP,
 
 		MasterKubeConfig: admin.DefaultNodeKubeConfigFile(args.CertArgs.CertDir, args.NodeName),
+
+		PodManifestConfig: &configapi.PodManifestConfig{
```
Comment (liggitt) [inline pkg/cmd/server/start/node_args.go]: Make podManifestConfig nil if optional, not specified with empty values inside

**Comment by derekwaynecarr** [comment]
Do we plan on using this to bootstrap an origin master server similar to how the default Kube setup is now running the master components of Kube in a pod?

File: pkg/cmd/server/start/node_args.go (hunk 12 above)
Comment (abhgupta) [inline pkg/cmd/server/start/node_args.go]: Do we actually need
```
PodManifestConfig: nil
```
Or should we ignore (not specify) the PodManifestConfig attribute?

**Comment by abhgupta** [comment]
@derekwaynecarr while there may be other use cases, the immediate use case is to allow the creation of a pod (specifically, fluentd) on every node by placing a pod manifest file on the nodes.

File: pkg/cmd/server/start/node_args.go (hunk 12 above)
Comment (liggitt) [inline pkg/cmd/server/start/node_args.go]: Either is fine, same end result. Explicitly setting to nil makes it clear it was intentionally omitted by default

File: pkg/cmd/server/api/types.go [hunk 13]
```diff
@@ -382,3 +386,13 @@ type CertInfo struct {
...
 }
+
+type PodManifestConfig struct {
+	// Path specifies the path for the pod manifest file or directory
+	// If its a directory, its expected to contain on or more manifest files
+	// This is used by the Kubelet to create pods on the node
+	Path string `json:"path"`
+	// FileCheckIntervalSeconds is the interval in seconds for checking the manifest file(s) for new data
+	// The interval needs to be a positive value
+	FileCheckIntervalSeconds int64 `json:"fileCheckIntervalSeconds"`
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: no JSON tags

**Comment by abhgupta** [comment]
@liggitt final review/merge please

**Comment by liggitt** [comment]
LGTM

**Comment by abhgupta** [comment]
[merge]

**Pull request merged**